router/src/completion.rs

```diff
@@ -0,0 +1,565 @@
+    pub text: String,
+    #[schema(example = "length")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub finish_reason: Option<FinishReason>,
+    // pub generated_tokens: u32,
+    // logprobs is not implemented, send None
+    pub logprobs: Option<Vec<u32>>,
+    #[schema(example = 0)]
+    pub index: u32,
+}
+
+#[derive(Serialize, ToSchema)]
+pub(crate) struct CompletionsResponse {
+    #[schema(example = "cmpl-abcdefgehij1234")]
+    pub id: String,
+    #[schema(example = "text_completion")]
+    pub object: String,
+    #[schema(example = 1589478379)]
+    pub created: u32,
+    #[schema(example = "tgi")]
+    pub model: String,
+    pub choices: Vec<CompletionChoices>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub usage: Option<Usage> ,
+}
+
+#[derive(Serialize, ToSchema)]
+pub(crate) struct ChatCompletionChoices {
+    #[schema(example = "test")]
+    pub message: ChatMessage,
+    #[schema(example = "length")]
+    pub finish_reason: Option<FinishReason>,
+    // pub generated_tokens: u32,
+    #[schema(example = 0)]
+    pub index: u32,
+}
+
+#[derive(Serialize, ToSchema)]
+pub(crate) struct ChatCompletionDeltaStreamChoices {
+    #[schema(example = "test")]
+    pub delta: ChatDeltaStreamMessage,
+    #[schema(example = "length")]
+    pub finish_reason: Option<FinishReason>,
+    // pub generated_tokens: u32,
+    #[schema(example = 0)]
+    pub index: u32,
+}
+
+#[derive(Serialize, ToSchema)]
+pub(crate) struct ChatCompletionsResponse {
+    #[schema(example = "chatcmpl-abcdefgehij1234")]
+    pub id: String,
+    #[schema(example = "chat.completion")]
+    pub object: String,
+    #[schema(example = 1589478380)]
+    pub created: u32,
+    #[schema(example = "tgi")]
+    pub model: String,
+    pub choices: Vec<ChatCompletionChoices>,
+    pub usage: Usage,
+}
+
+#[derive(Serialize, ToSchema)]
+pub(crate) struct ChatCompletionsStreamResponse {
+    #[schema(example = "chatcmpl-abcdefgehij1234")]
+    pub id: String,
+    #[schema(example = "chat.completion.chunk")]
+    pub object: String,
+    #[schema(example = 1589478380)]
+    pub created: u32,
+    #[schema(example = "tgi")]
+    pub model: String,
+    pub choices: Vec<ChatCompletionDeltaStreamChoices>,
+}
+
+pub const CHAT_USER_PRE: &'static str = match option_env!("TGICHAT_USER_PRE") {
+    Some(v) => v,
+    None => "",
+};
+pub const CHAT_USER_POST: &'static str = match option_env!("TGICHAT_USER_POST") {
+    Some(v) => v,
+    None => "",
+};
+pub const CHAT_ASS_PRE: &'static str = match option_env!("TGICHAT_ASS_PRE") {
+    Some(v) => v,
+    None => "",
+};
+pub const CHAT_ASS_POST:&'static str = match option_env!("TGICHAT_ASS_POST") {
+    Some(v) => v,
+    None => "",
+};
+pub const CHAT_SYS_PRE: &'static str = match option_env!("TGICHAT_SYS_PRE") {
+    Some(v) => v,
+    None => "",
+};
+pub const CHAT_SYS_POST: &'static str = match option_env!("TGICHAT_SYS_POST") {
+    Some(v) => v,
+    None => "",
+};
```

Comment:
What if we want to inject a special token, for example `<s>` to the inputs?

Reply:
I updated the code for this section, you are correct, this input type is `String`, not `Token`. In many models, e.g. LLama-2, there is no single special token for 'sys', 'user', 'assistant' in the tokenizer. As this rust code at this stage operates on string level, even if the sys prompt is made out of two special tokens in the tokenizer, in the `router` you cannot perform such a string sanitation, as e.g. `<<SYS>>[INST]` is two tokens.

You could perhaps set a feature flag to not allow any sequence from the 6 variables in the user message - that is possible, but lets not overload this Pull request.

Example for LLama-2:
```bash
export TGICHAT_SYS_PRE="<<SYS>>[INST]"
export TGICHAT_SYS_POST="[\INST]<<SYS>>"
text-generation-launcher --model-id meta-llama/Llama-2-13b-chat
```

Reply:
Understood. I think this is a limitation in the original protobuf protocol, where the `inputs` is a string. Ideally, `inputs` should be a ChatML. I agree it's not an issue that needed to be addressed in this PR.